File: server/controllers/ports.py
# get connected serial ports
import serial
import serial.tools.list_ports
import time
from flask_cors import cross_origin
from sqlalchemy.exc import SQLAlchemyError
from flask import Blueprint, jsonify, request, make_response
from models.printers import Printer

ports_bp = Blueprint("ports", __name__)

"""
    Get all connected serial ports. Only return ports that are connected to 3D printers. 
"""
@ports_bp.route("/getports",  methods=["GET"])
def getPorts():
    printerList = Printer.getConnectedPorts()
    return jsonify(printerList)

"""
    Gets all registered printers from the database
"""
@ports_bp.route("/getprinters", methods=["GET"])
def getRegisteredPrinters():  
    try: 
        res = Printer.get_registered_printers()
        return res
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500

"""
    Register a new printer in the database and create a thread for it. App imported directly in the function 
    to avoid circular import error.
"""
@ports_bp.route("/register", methods=["POST"])
def registerPrinter(): 
    try: 
        from app import printer_status_service
        data = request.get_json() 
        device = data['printer']['device']
        description = data['printer']['description']
        hwid = data['printer']['hwid']
        name = data['printer']['name']
        
        res = Printer.create_printer(device=device, description=description, hwid=hwid, name=name, status='ready')
        if(res["success"] == True):
            id = res['printer_id']
            thread_data = {
                "id": id, 
                "device": device,
                "description": description,
                "hwid": hwid,
                "name": name
            }
            
            printer_status_service.create_printer_threads([thread_data])
        
        return res
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500

"""
    Deletes printer from database. The thread is removed through a controller in statusService.py 
"""
@ports_bp.route("/deleteprinter", methods=["POST"])
def delete_printer():
    try: 
        data = request.get_json()
        printerid = data['printerid']
        res = Printer.deletePrinter(printerid)
        return res 
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500
    
"""
    Edits name of printer in database. Name edited in-memory in statusService.py
"""
@ports_bp.route("/editname", methods=["POST"])
def edit_name(): 
    try: 
        data = request.get_json() 
        printerid = data['printerid']
        name = data['name']
        res = Printer.editName(printerid, name)
        return res 
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500
    
"""
    Tells user if printer is connected to a serial port or not and if the port it is registered under 
    is the port that the printer is currently connected to.
"""
@ports_bp.route("/diagnose", methods=["POST"])
def diagnose_printer():
    try:
        data = request.get_json() 
        device = data['device']
        res = Printer.diagnosePrinter(device)
        return res
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500
    
"""
    Route to send the "printer home" gcode command while the printer is registering a printer, so the user can tell 
    which printer they have selected. 
"""
@ports_bp.route("/movehead", methods=["POST"])
def moveHead():
    try: 
        data = request.get_json()
        port = data['port']
        
        res = Printer.moveHead(port)
        if res == "none": 
            return {"success": False, "message": "Head move unsuccessful."}
        
        return {"success": True, "message": "Head move successful."}
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500

"""
    Method to change the order of printer threads so the user can rearrange the order of printers on the UI (Main view)
"""
@ports_bp.route("/moveprinterlist", methods=["POST"])
def movePrinterList():
    try:
        from app import printer_status_service
        data = request.get_json()
        printersIds = data['printersIds']

        res = printer_status_service.movePrinterList(printersIds) 
        if res == "none": 
            return {"success": False, "message": "Printer list not updated."}
               
        return jsonify({"success": True, "message": "Printer list successfully updated."})
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500

File: client/src/model/ports.ts
import { useRouter } from 'vue-router'
import { ref, computed, onUnmounted } from 'vue'
import * as myFetch from './myFetch'
import { toast } from './toast'
import { type Job } from './jobs'
import { socket } from './myFetch'

export function api(action: string, body?: unknown, method?: string, headers?: any) {
  headers = headers ?? {}
  return myFetch.api(`${action}`, body, method, headers).catch((err) => console.log(err))
}

export interface Device {
  device: string
  description: string
  hwid: string
  name?: string
  status?: string
  date?: Date
  id?: number
  error?: string
  canPause?: number
  queue?: Job[] //  Store job array to store queue for each printer.
  isQueueExpanded?: boolean
  isInfoExpanded?: boolean
  extruder_temp?: number
  bed_temp?: number
  colorChangeBuffer?: number
}

/*
  Global variable to store the list of printers and printer queues based on in-memory thread data from backend 
*/
export let printers = ref<Device[]>([])

/*
  Get all connected serial ports. Only return ports that are connected to 3D printers. 
*/
export function useGetPorts() {
  return {
    async ports() {
      try {
        const response = await api('getports')
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  Register printer based on data collected from useGetPorts 
*/
export function useRegisterPrinter() {
  return {
    async register(printer: Device) {
      try {
        const response = await api('register', { printer })
        if (response) {
          if (response.success === false) {
            toast.error(response.message)
          } else if (response.success === true) {
            toast.success(response.message)
          } else {
            console.error('Unexpected response:', response)
            toast.error('Failed to register printer. Unexpected response')
          }
        } else {
          console.error('Response is undefined or null')
          toast.error('Failed to register printer. Unexpected response')
        }
      } catch (error) {
        console.error(error)
        toast.error('An error occurred while registering the printer')
      }
    }
  }
}

/*
  Gets all registered printers from the database to display on Registered View 
*/
export function useRetrievePrinters() {
  return {
    async retrieve() {
      try {
        const response = await api('getprinters')
        return response.printers
      } catch (error) {
        console.error(error)
      }
    }
  }
}


/*
  Main function to retrieve thread data from backend and store it in the printers variable.
*/
export function useRetrievePrintersInfo() {
  return {
    async retrieveInfo() {
      try {
        const response = await api('getprinterinfo')
        return response // return the response directly
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  Set status of printer 
*/
export function useSetStatus() {
  return {
    async setStatus(printerid: number | undefined, status: string) {
      try {
        const response = await api('setstatus', { printerid, status })
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  Delete and recreate thread data for printer
*/
export function useHardReset() {
  return {
    async hardReset(printerid: number | undefined) {
      try {
        const response = await api('hardreset', { printerid })
        if (response) {
          if (response.success == false) {
            toast.error(response.message)
          } else if (response.success === true) {
            toast.success(response.message)
          } else {
            console.error('Unexpected response:', response)
            toast.error('Failed to release job. Unexpected response.')
          }
        } else {
          console.error('Response is undefined or null')
          toast.error('Failed to release job. Unexpected response')
        }
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  If a printer is deregistered, set all FKs of jobs connected to that printer to null
*/
export function useNullifyJobs() {
  return {
    async nullifyJobs(printerid: number | undefined) {
      try {
        const response = await api('nullifyjobs', { printerid })
        if (response) {
          if (response.success == false) {
            toast.error(response.message)
          } else if (response.success === true) {
            toast.success(response.message)
          } else {
            console.error('Unexpected response:', response)
            toast.error('Failed to nullify jobs. Unexpected response.')
          }
        } else {
          console.error('Response is undefined or null')
          toast.error('Failed to nullify jobs. Unexpected response')
        }
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  Delete printer from database
*/
export function useDeletePrinter() {
  return {
    async deletePrinter(printerid: number | undefined) {
      try {
        const response = await api('deleteprinter', { printerid })
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  Delete in-memory thread data for printer
*/
export function useRemoveThread() {
  return {
    async removeThread(printerid: number | undefined) {
      try {
        const response = await api('removethread', { printerid })
        if (response) {
          if (response.success == false) {
            toast.error(response.message)
          } else if (response.success === true) {
            toast.success(response.message)
          } else {
            console.error('Unexpected response:', response)
            toast.error('Failed to remove thread. Unexpected response.')
          }
        } else {
          console.error('Response is undefined or null')
          toast.error('Failed to remove thread. Unexpected response')
        }
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  Edit name of printer in database 
*/
export function useEditName() {
  return {
    async editName(printerid: number | undefined, name: string) {
      try {
        const response = await api('editname', { printerid, name })
        if (response) {
          if (response.success == false) {
            toast.error(response.message)
          } else if (response.success === true) {
            toast.success(response.message)
          } else {
            console.error('Unexpected response:', response)
            toast.error('Failed to edit name. Unexpected response.')
          }
        } else {
          console.error('Response is undefined or null')
          toast.error('Failed to edit name. Unexpected response')
        }
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  Edit name of printer in thread data
*/
export function useEditThread() {
  return {
    async editThread(printerid: number | undefined, newname: string) {
      try {
        const response = await api('editNameInThread', { printerid, newname })
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  Tells user if printer is connected to a serial port or not and if the port it is registered under 
  is the port that the printer is currently connected to.
*/
export function useDiagnosePrinter() {
  return {
    async diagnose(device: string) {
      try {
        const response = await api('diagnose', { device })
        if (response) {
          if (response.success == false) {
            toast.error(response.message)
          } else if (response.success === true) {
            toast.success(response.message)
          } else {
            console.error('Unexpected response:', response)
            toast.error('Failed to diagnose printer. Unexpected response.')
          }
        } else {
          console.error('Response is undefined or null')
          toast.error('Failed to diagnose printer. Unexpected response')
        }
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
    Lists all of the ports currently connected to the machine. If the port the printer is registered under does not match the  
    port the printer is currently connected to, update the port in the database. We check this by comparing the hwid 
    returned by the connected port and the hwid stored in the database. 
*/
export function useRepair() {
  return {
    async repair() {
      try {
        const response = await api('repairports')
        if (response) {
          if (response.success == false) {
            toast.error(response.message)
          } else if (response.success === true) {
            toast.success(response.message)
          } else {
            console.error('Unexpected response:', response)
            toast.error('Failed to repair ports. Unexpected response.')
          }
        } else {
          console.error('Response is undefined or null')
          toast.error('Failed to repair ports. Unexpected response')
        }
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
    Route to send the "printer home" gcode command while the printer is registering a printer, so the user can tell 
    which printer they have selected. 
*/
export function useMoveHead() {
  return {
    async move(port: string) {
      try {
        const response = await api('movehead', { port })
        if (response) {
          if (response.success == false) {
            toast.error(response.message)
          } else if (response.success === true) {
            toast.success(response.message)
          } else {
            console.error('Unexpected response:', response)
            toast.error('Failed to move head. Unexpected response.')
          }
        } else {
          console.error('Response is undefined or null')
          toast.error('Failed to move head. Unexpected response')
        }
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}

/*
  Method to change the order of printer threads so the user can rearrange the order of printers on the UI (Main view)
*/
export function useMovePrinterList() {
  return {
    async movePrinterList(printers: Device[]) {
      try {
        // make new array of printer id's in the order they are in the printers array
        const printersIds = printers.map((printer) => printer.id)
        const response = await api('moveprinterlist', { printersIds })
        return response
      } catch (error) {
        console.error(error)
      }
    }
  }
}
